File: leetcode_src/src/Candy.py
from typing import *

class Solution:
    def candy(self, ratings: List[int]) -> int:
        n = len(ratings)
        left = [0] * n

        left[0] = 1
        for i in range(1, n):
            left[i] = left[i-1]+1 if ratings[i] > ratings[i-1] else 1

        right = 1
        res = max(right, left[n-1])
        for j in range(n-2, -1, -1):
            right = right+1 if ratings[j] > ratings[j+1] else 1
            res += max(right, left[j])

        return res

print(Solution().candy([2,0,1]))

File: leetcode_src/src/Candy.java
public class Candy {
    // two passes
    public int candy(int[] ratings) {
        int[] left = new int[ratings.length];
        int[] right = new int[ratings.length];

        left[0] = 1;
        right[ratings.length-1] = 1;

        for (int i = 1; i < ratings.length; i++) {
            left[i] = (ratings[i] > ratings[i - 1]) ? left[i - 1] + 1 : 1;
        }
        for (int j = ratings.length-2; j >= 0; j--) {
            right[j] = (ratings[j] > ratings[j + 1]) ? right[j + 1] + 1 : 1;
        }

        int res = 0;
        for(int k = 0; k < ratings.length; k++) {
            res += Math.max(left[k], right[k]);
        }

        return res;
    }

    // greedy algorithm
    public int candy_greedy(int[] ratings) {
        int res = 0;
        int candy = 1;

        res = 1;
        for(int i = 1; i < ratings.length; i++) {
            if(ratings[i] > ratings[i-1]) {
                candy++;
                res += candy;
            }
            else {

            }
        }

        return -1;
    }

    public static void main(String[] args) {
        Candy sol = new Candy();

        System.out.println(sol.candy(new int[]{1,0,2}));
    }
}
